File: src/modules/httpsrc/python/testhttpsrc.py
"""@package testnet
   @file testnet.py
   @author Ivan Murashko <ivan.murashko@gmail.com>

   Copyright (c) 2007-2010 Kalinka Team

   This file is part of Kalinka mediaserver.

   Permission is hereby granted, free of charge, to any person obtaining
   a copy of this software and associated documentation files (the
   "Software"), to deal in the Software without restriction, including
   without limitation the rights to use, copy, modify, merge, publish,
   distribute, sublicense, and/or sell copies of the Software, and to
   permit persons to whom the Software is furnished to do so, subject to
   the following conditions:

   The above copyright notice and this permission notice shall be included
   in all copies or substantial portions of the Software.

   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
   EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
   MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
   IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
   CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
   TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
   SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

   CHANGE HISTORY

   @date
   - 2010/12/04 created by ipp (Ivan Murashko)
"""

import klk.httpsrc
import klk.server

import unittest
import os
import sys


class HTTPSourceTest(unittest.TestCase):
    """ HTTPSourcework module unit test
    """

    def setUp(self):
        """ Creates test data
        """

        # Factory object
        config_file = os.path.join(klk.paths.INSTALL_PATH, \
                                       'etc/klk/klktest.conf')
        config = klk.config.Config(config_file)
        self._factory = klk.Factory()
        self._factory.config = config

        db = klk.db.DB(config)
        db.recreateDB()
        # fill DB with test data
        db.executeScript('testhttpsrc.sql')

    def testHTTPSources(self):
        """ HTTPSource module: HTTP sources info test
        """

        # initially we have 2 sources
        server = self._factory.servers[self._factory.config.server]
        self.assert_(server)        
        module = server.modules[klk.httpsrc.NAME]
        sources = module.sources
        self.assert_(sources)        
        self.assert_(len(sources) == 2)

        # add several sources
        module.addSource('source1', 'http://webcam1/', 'login', '12345' )
        module.addSource('source2', 'http://webcam2/', '', '' )

        # test the result
        sources = module.sources
        self.assert_(sources)        
        self.assert_(len(sources) == 4)
        self.assert_('source1' in sources.keys())
        self.assert_('source2' in sources.keys())

        item = sources['source1']
        self.assert_(item.name == 'source1')
        self.assert_(item.address == 'http://webcam1/')
        self.assert_(item.login == 'login')
        self.assert_(item.passwd == '12345')

        item = sources['source2']
        self.assert_(item.name == 'source2')
        self.assert_(item.address == 'http://webcam2/')
        self.assert_(item.login == '')
        self.assert_(item.passwd == '')

        # check the del
        module.delSource('source1')
        sources = module.sources
        self.assert_(sources)        
        self.assert_(len(sources) == 3)
        self.assert_('source1' not in sources.keys())
        self.assert_('source2' in sources.keys())

        module.delSource('source2')
        sources = module.sources
        self.assert_(sources)        
        self.assert_(len(sources) == 2)
        self.assert_('source1' not in sources.keys())
        self.assert_('source2' not in sources.keys())
        

File: src/modules/httpsrc/testhttpsrc.cpp
/**
   @file testhttpsrc.cpp
   @brief This file is part of Kalinka mediaserver.
   @author Ivan Murashko <ivan.murashko@gmail.com>

   Copyright (c) 2007-2012 Kalinka Team

   Permission is hereby granted, free of charge, to any person obtaining
   a copy of this software and associated documentation files (the
   "Software"), to deal in the Software without restriction, including
   without limitation the rights to use, copy, modify, merge, publish,
   distribute, sublicense, and/or sell copies of the Software, and to
   permit persons to whom the Software is furnished to do so, subject to
   the following conditions:

   The above copyright notice and this permission notice shall be included
   in all copies or substantial portions of the Software.

   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
   EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
   MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
   IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
   CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
   TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
   SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

   CHANGE HISTORY

   @date
   - 2009/11/21 created by ipp (Ivan Murashko)
   - 2010/01/06 header was changed by header.py script
   - 2011/01/01 header was changed by header.py script
   - 2012/02/03 header was changed by header.py script
*/

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include "testhttpsrc.h"
#include "defines.h"
#include "testid.h"
#include "httpsrc.h"
#include "testutils.h"
#include "testcli.h"
#include "testhelpmodule.h"


using namespace klk;
using namespace klk::httpsrc;

// The function inits module's unit test
void klk_module_test_init()
{
    CPPUNIT_TEST_SUITE_NAMED_REGISTRATION(TestHTTPSrc,
                                          MODNAME);
    CPPUNIT_TEST_SUITE_NAMED_REGISTRATION(TestCLI,
                                          MODNAME);
    CPPUNIT_REGISTRY_ADD(MODNAME, test::ALL);
}

//
// TestHTTPSrc class
//


// Constructor
TestHTTPSrc::TestHTTPSrc() :
    test::TestModuleWithDB("testhttpsrc.sql")
{
}

// Loads all necessary modules at setUp()
void TestHTTPSrc::loadModules()
{
    CPPUNIT_ASSERT(m_modfactory);
    // now load the it
    m_modfactory->load(MODID);

    // add helper module
    this->addDebug<TestHelpModule>();
    // load helper module
    m_modfactory->load(test::HELP_MOD_ID);
}


// The unit test itself
void TestHTTPSrc::test()
{
    test::printOut("\nHTTP source test ... ");

    // give some time to module load
    sleep(3);

    // start the processing
    IMessagePtr start_msg = m_msgfactory->getMessage(test::HELP_MSG_ID);
    CPPUNIT_ASSERT(start_msg);
    m_modfactory->sendMessage(start_msg);

    sleep(10);

    Result rc = getModule<TestHelpModule>(test::HELP_MOD_ID)->getResult();
    CPPUNIT_ASSERT(rc == OK);
}

